### ISSUE TITLE
Judge's feedback

### ISSUE THREAD

**Issue by ghost** [Billz2me/railschallenge-city-watch#1]
Placeholder issue for judge's feedback.

**Comment by ghost** [comment]
Models and DB:
- Interesting choice to explicitly disable STI. Model-specific behavior isn't required, for sure, but are empty classes (the decedent classes) better than explicitly disabling the `type` column?

General:
- You can leave the comments off of the controller actions if you like. They're standard actions, and they just add noise to the file. Leaving the comments in place **won't** count against your score or anything. This is just an FYI.

`app/models/concerns` folder
- Not sure these concerns really need to be defined. While it’s true that using concerns is a way to cut down on the code in the model, the overarching goal of them is to encourage code reuse. Since the concerns you’ve defined are only used on their respective models, there’s not much motivation for using concerns in this case.

`config/routes`:
- Design choices elsewhere in the application (how to handle 404s, etc.) has resulted in additional complexity in this file. Might want to reconsider this part of the design.

EmergenciesController:
- Line 14: `@emergency.valid? && @emergency.save` is redundant - if it's valid it should save
- Line 24: Good use of a single line to do the rendering

**Comment by railschallenge-judge01** [comment]
General
- You don't need to use class variables in the many of the controller actions
  - class variables are needed when you are rendering views
  - This project renders inside the action, so '@responder' and '@emergency' variables can all be 'responder' and 'emergency'
- permit_parameters! method
  - explicitly call out `when :update` rather than `else`
  - otherwise, expanded use of this method could caused other actions to use the else param requrements
- You probably don't need to memoize `@permitted_parameters`
  - Looks like it's only getting called once per request
- Good use of method documentation (looks like tomdoc format)

Responder Model
- `self.primary_key = :name` is a dangerous choice
  - it cleans up some code in the show/delete controller actions
  - but names can change over time, making it an unreliable PK

Emergency Model
- `self.primary_key = :code` is less dangerous than :name in the Responder model
  - Codes are less likely to change than names
  - However, it's probably better practice to consistely use the id for the PK

ResponderCapacityReport
- Nice use of concerns to keep the models thin
- `Responder::TYPES.each_with_object({})` can be `each.with_object`
  - with_object is more flexible, as it can be applied to any enumerator
  - Either that or much the same can be accomplished with `Responder::TYPES.inject...`

**Comment by Billz2me** [comment]
General
- You don't need to use class variables in the many of the controller actions
  - class variables are needed when you are rendering views
  - This project renders inside the action, so '@responder' and '@emergency' variables can all be 'responder' and 'emergency'
- permit_parameters! method
  - explicitly call out `when :update` rather than `else`
  - otherwise, expanded use of this method could caused other actions to use the else param requrements
- You probably don't need to memoize `@permitted_parameters`
  - Looks like it's only getting called once per request
- Good use of method documentation (looks like tomdoc format)

**Billz2me** - Will change the controllers to not use instance variables. edit: On second thought, I actually like the syntax highlighting which puts emphasis on the object being acted upon. How do you feel about this? 
**Billz2me** - Will refactor permit_parameters! 
**Billz2me** - Good point, will remove memoization.  

Responder Model
- `self.primary_key = :name` is a dangerous choice
  - it cleans up some code in the show/delete controller actions
  - but names can change over time, making it an unreliable PK

**Billz2me** - There is an explicitly requirement that name can't be changed, but the point is valid and I can change it. 

Emergency Model
- `self.primary_key = :code` is less dangerous than :name in the Responder model
  - Codes are less likely to change than names
  - However, it's probably better practice to consistely use the id for the PK

**Billz2me** - I could go either way on this one, I think that code is the natural PK for the Emergency, so I decided to omit the vestigial "ID". Since "Responder" needs to refer to "emergency_code", I felt as though that cemented it even further as a reference key. Otherwise I'd have to do an additional query like @responder.emergency.code if my relationship was defined on an integer ID.

ResponderCapacityReport
- Nice use of concerns to keep the models thin
- `Responder::TYPES.each_with_object({})` can be `each.with_object`
  - with_object is more flexible, as it can be applied to any enumerator
  - Either that or much the same can be accomplished with `Responder::TYPES.inject...`

**Billz2me** - I actually started with .inject, but Rubocop suggested that I use each_with_object.

**Comment by Billz2me** [comment]
Models and DB:
- Interesting choice to explicitly disable STI. Model-specific behavior isn't required, for sure, but are empty classes (the decedent classes) better than explicitly disabling the `type` column?

**Billz2me** - I do think that disabling STI here is the right choice. Perhaps an abstraction would necessary if there were further requirements that were responder type specific. I think that empty classes in this case would just be unnecessary bloat.

General:
- You can leave the comments off of the controller actions if you like. They're standard actions, and they just add noise to the file. Leaving the comments in place **won't** count against your score or anything. This is just an FYI.

**Billz2me** - Thanks for the heads up. 

`app/models/concerns` folder
- Not sure these concerns really need to be defined. While it’s true that using concerns is a way to cut down on the code in the model, the overarching goal of them is to encourage code reuse. Since the concerns you’ve defined are only used on their respective models, there’s not much motivation for using concerns in this case.

**Billz2me** - Thanks for the comment, In my opinion a concern doesn't necessarily have to be reused. Take rails source code for example, most models are composed of many concerns that are only used once, but it is a way to easily contain logic "concerning" a single logical concept. I split it out into a concern because it is solely "concerned" with one concept - dispatching. 

`config/routes`:
- Design choices elsewhere in the application (how to handle 404s, etc.) has resulted in additional complexity in this file. Might want to reconsider this part of the design.

**Billz2me** - Can you please elaborate? What is the additional complexity? If you're referring to the "except" clause, its there because defining routes that aren't intended to be used is generally bad practice. 

EmergenciesController:
- Line 14: `@emergency.valid? && @emergency.save` is redundant - if it's valid it should save
- Line 24: Good use of a single line to do the rendering

**Billz2me** - Thanks, will remove the redundant validation.

**Comment by ghost** [comment]
Hm, never mind on the routes complexity comment.

> Billz2me - There is an explicitly requirement that name can't be changed, but the point is valid and I can change it.

It's true that requirement is in the spec, but that reason alone doesn't necessarily make it a good candidate for a PK. Schemas can change. `name` and `code` are convenient lookup values for the API, but that decision was primarily made to hide internal DB `id` columns from the API spec. I don't know if I'd assume a non-changing name/code field means it should be a PK. You can still use internal `id` fields for association linking.

Regarding the rest of your comments: fair enough.

This will be the last feedback I can provide before the end of the submission period, as we have many more repos for which to provide early reviews. But I wanted to say thanks for participating the challenge!

**Comment by Billz2me** [comment]
Agreed with the PK comment on Responder name, I've made that change. I think :code is a natural PK for the Emergency however. I don't see a problem with using a non-changing code as a PK actually, as long as you adhere to that non-changing contract. Many databases don't have an "ID" column, thats kind of a Rails concept. In lieu of such a column, the natural PK is used. In this case it actually improves performance because I can get responder.emergency_code from a column and not by doing a separate query for the responder.emergency.code. 

Thanks for your review!

**Comment by ghost** [comment]
You're correct about the performance, and you're correct about the `id` thing being a Rails convention. Simply have to draw the line somewhere.

Thanks for the conversation. The judges are going to have to move on, spending the remaining time on early code reviews for other forks that haven't yet received their first attention.

Best of luck in the final scoring!

**Comment by ghost** [comment]
Verified score of `114 points` on the automated scoring, based on the current master branch.

**Comment by ghost** [comment]
:trophy: Congrats on [placing 3rd in the first **rails**challenge](http://railschallenge.com/results/)!

You had the best `capacity` component according to the judges. We'd like to invite you to write up something about your approach to the capacity code, as well as your overall approach, to be included in the feedback for all participants.

Please reach out to us at [info@railschallenge.com](mailto:info@railschallenge.com)

**Comment by Billz2me** [comment]
Great news! So glad to hear I placed third. I'm on vacation right now so I may not be able to get back to you right away with a full response. When would you like this by?

**Comment by ghost** [comment]
Regarding the write-up: I think I responded via email as well, but take your time with the write up. We'll post something probably in a week or so. There's a lot of analysis, and all of the top 3 placers were asked to write something, so it make take a while to get everything back from everyone.

Also, since you got the best `capacity` component, focus your write-up on that, but also talk about your overall approach to the problem.

Thanks again!

**Comment by ghost** [comment]
Your individual score breakdown:

279 points total:
- 114 for passing tests
- 165 for the judges' scoring
